analyzers/tests/SonarAnalyzer.UnitTest/Extensions/DiagnosticDescriptorExtensionsTest.cs

```diff
@@ -0,0 +1,108 @@
+    [DataRow(false, ProjectType.Test, MainTag)]
+    [DataRow(false, ProjectType.Test, MainTag, MainTag)]
+    public void HasMatchingScope_SingleDiagnostic_WithOneOrMoreScopes_SonarLint(bool expectedResult, ProjectType projectType, params string[] ruleTags)
+    {
+        var compilation = new SnippetCompiler("// Nothing to see here").SemanticModel.Compilation;
+        var diagnostic = TestHelper.CreateDescriptor("Sxxx", ruleTags);
+        diagnostic.HasMatchingScope(compilation, projectType == ProjectType.Test, false).Should().Be(expectedResult);
+    }
+
+    [DataTestMethod]
+    [DataRow(true, ProjectType.Product, MainTag)]
+    [DataRow(true, ProjectType.Product, MainTag, UtilityTag)]
+    [DataRow(true, ProjectType.Product, MainTag, TestTag)]
+    [DataRow(true, ProjectType.Product, MainTag, TestTag, UtilityTag)]
+    [DataRow(true, ProjectType.Test, TestTag)]
+    [DataRow(true, ProjectType.Test, TestTag, UtilityTag)]
+    [DataRow(true, ProjectType.Test, MainTag, TestTag, UtilityTag)]     // Utility rules with scope Test&Main do run on test code under scanner context.
+    [DataRow(false, ProjectType.Test, MainTag, TestTag)]                // Rules with scope Test&Main do not run on test code under scanner context for now.
+    [DataRow(false, ProjectType.Product, TestTag)]
+    [DataRow(false, ProjectType.Product, TestTag, UtilityTag)]
+    [DataRow(false, ProjectType.Product, TestTag, TestTag)]
+    [DataRow(false, ProjectType.Test, MainTag)]
+    [DataRow(false, ProjectType.Test, MainTag, UtilityTag)]
+    [DataRow(false, ProjectType.Test, MainTag, MainTag)]
+    public void HasMatchingScope_SingleDiagnostic_WithOneOrMoreScopes_Scanner(bool expectedResult, ProjectType projectType, params string[] ruleTags)
+    {
+        var compilation = new SnippetCompiler("// Nothing to see here").SemanticModel.Compilation;
+        var diagnostic = TestHelper.CreateDescriptor("Sxxx", ruleTags);
+        diagnostic.HasMatchingScope(compilation, projectType == ProjectType.Test, true).Should().Be(expectedResult);
+    }
+
+    [DataTestMethod]
+    [DataRow(true, ProjectType.Product, MainTag, MainTag)]
+    [DataRow(true, ProjectType.Product, MainTag, MainTag)]
+    [DataRow(true, ProjectType.Product, MainTag, TestTag)]
+    [DataRow(true, ProjectType.Test, TestTag, TestTag)]
+    [DataRow(true, ProjectType.Test, TestTag, MainTag)]
+    [DataRow(false, ProjectType.Product, TestTag, TestTag)]
+    [DataRow(false, ProjectType.Test, MainTag, MainTag)]
+    public void HasMatchingScope_MultipleDiagnostics_WithSingleScope_SonarLint(bool expectedResult, ProjectType projectType, params string[] rulesTag)
+    {
+        var compilation = new SnippetCompiler("// Nothing to see here").SemanticModel.Compilation;
+        var diagnostics = rulesTag.Select(x => TestHelper.CreateDescriptor("Sxxx", x));
+        diagnostics.Any(x => x.HasMatchingScope(compilation, projectType == ProjectType.Test, false)).Should().Be(expectedResult);
```

Comment:
This and the following test doesn't make much sense, as they don't test the logic but rather replication of the logic. It is refactored in PR 15